async-book/src/execution/future.md

```diff
@@ -0,0 +1,178 @@
+make more progress.
+
+For example, a simplified `SocketRead` `Future` might look like this:
+
+```rust
+struct SocketRead<'a> {
+    socket: &'a Socket,
+}
+
+impl SimpleFuture for SocketRead<'_> {
+    type Output = Vec<u8>;
+
+    fn poll(&mut self, wake: fn()) -> Poll<Self::Output> {
+        if self.socket.has_data_to_read() {
+            // The socket has data-- read it into a buffer and return it.
+            Poll::Ready(self.socket.read_buf())
+        } else {
+            // The socket does not yet have data.
+            //
+            // Arrange for `wake` to be called once data is available.
+            // When data becomes available, `wake` will be called, and the
+            // user of this `Future` will know to call `poll` again and
+            // receive data.
+            self.socket.set_readable_callback(wake);
+        }
+    }
+}
+```
+
+This model of `Future`s allows for composing together multiple asynchronous
+operations without needing intermediate allocations. Running multiple futures
+at once or chaining futures together can be implemented via allocation-free
+state machines, like this:
+
+```rust
+/// A SimpleFuture that runs two other futures to completion concurrently.
+///
+/// Concurrency is acheived via the fact that calls to `poll` each future
+/// may be interleaved, allowing each future to advance itself at its own pace.
+struct Join2 {
+    // Each field may contain a future that should be run to completion.
+    // If the future has already completed, the field is set to `None`.
+    a: Option<FutureA>,
+    b: Option<FutureB>,
+}
+
+impl SimpleFuture for Join2 {
+    type Output = ();
+    fn poll(&mut self, wake: fn()) -> Poll<Self::Output> {
+        // Attempt to complete future `a`.
+        let finished_a = match &mut self.a {
+            Some(a) => {
+                match a.poll(wake) {
+                    Poll::Ready(()) => true,
+                    Poll::Pending => false,
+                }
+            }
+            None => true,
+        };
+        if finished_a { self.a.take() }
+
+        // Attempt to complete future `b`.
+        let finished_b = match &mut self.b {
+            Some(b) => {
+                match b.poll(wake) {
+                    Poll::Ready(()) => true,
+                    Poll::Pending => false,
+                }
+            }
+            None => true,
+        };
+        if finished_b { self.b.take() }
+
+        if finished_a && finished_b {
+            // Both futures have completed-- we can return successfully
+            Poll::Ready(())
+        } else {
+            // One or both futures still have work to do, and will call
+            // `wake()` when progress can be made.
+            Poll::Pending
+        }
+    }
+}
+```
+
+This shows how multiple futures can be run simultaneously without needing
+separately allocated tasks, allowing for more efficient asynchronous programs.
```

Comment:
> needing separately allocated tasks

What's a `task` here? This is the first time I came across a `task` in the book, is this different from a `Future`?

In general there's emphasis on how the model of the `Future` trait allows for allocation free asynchronous computing. Is there any way to show why it's better this way? I get that allocations are a bad thing when going for high performance but it would be good to mention a reason why such an emphasis on no allocation is placed. Just my 2 cents.

Reply:
I've added some text to clarify what a task is, and how they interact with wakeups. Let me know if it's unclear of if you have more questions!